File: packages/nexus-coordinator/src/nexus_coordinator/forge.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Multi-forge URL detection and helpers for deploy-from-repo.

Sprint 14 Phase A — supports GitHub, GitLab, Codeberg, and generic
Gitea instances. The coordinator uses these helpers to:

1. Detect the forge type from a repo URL.
2. Construct a raw-file URL to pre-check ``SBFB.json`` without cloning.
3. Check whether a repository is publicly accessible.
4. Normalize the clone URL (strip fragments, trailing slashes).
"""

from __future__ import annotations

import re
from enum import Enum
from typing import NamedTuple

import httpx
import structlog

_log = structlog.get_logger(__name__)


class ForgeType(str, Enum):
    GITHUB = "github"
    GITLAB = "gitlab"
    CODEBERG = "codeberg"
    GITEA = "gitea"
    UNKNOWN = "unknown"


class ForgeInfo(NamedTuple):
    forge: ForgeType
    owner: str
    repo: str
    host: str


# Patterns match the most common forge URLs.
# Order matters — Codeberg before generic Gitea.
_FORGE_PATTERNS: list[tuple[re.Pattern[str], ForgeType]] = [
    (re.compile(r"https?://github\.com/(?P<owner>[^/]+)/(?P<repo>[^/.]+)"), ForgeType.GITHUB),
    (re.compile(r"https?://gitlab\.com/(?P<owner>[^/]+)/(?P<repo>[^/.]+)"), ForgeType.GITLAB),
    (re.compile(r"https?://codeberg\.org/(?P<owner>[^/]+)/(?P<repo>[^/.]+)"), ForgeType.CODEBERG),
    # Generic Gitea: any host with /<owner>/<repo> path.
    # Must be last — it's a fallback for self-hosted instances.
    (re.compile(r"https?://(?P<host>[^/]+)/(?P<owner>[^/]+)/(?P<repo>[^/.]+)"), ForgeType.GITEA),
]


def detect_forge(repo_url: str) -> ForgeInfo:
    """Detect the forge type from a repository URL.

    Returns a :class:`ForgeInfo` with the parsed owner, repo, host,
    and forge type. Unknown forges return ``ForgeType.UNKNOWN`` with
    empty owner/repo fields.
    """
    url = normalize_clone_url(repo_url)
    for pattern, forge_type in _FORGE_PATTERNS:
        m = pattern.match(url)
        if m:
            owner = m.group("owner")
            repo = m.group("repo")
            host = m.group("host") if "host" in m.groupdict() else ""
            return ForgeInfo(forge=forge_type, owner=owner, repo=repo, host=host)
    return ForgeInfo(forge=ForgeType.UNKNOWN, owner="", repo="", host="")


def raw_file_url(repo_url: str, path: str, ref: str = "HEAD") -> str | None:
    """Construct a URL to fetch a raw file from the repo without cloning.

    Returns ``None`` if the forge type is unknown (generic Gitea
    instances don't have a standardized raw URL pattern that works
    without API tokens).
    """
    info = detect_forge(repo_url)

    if info.forge == ForgeType.GITHUB:
        return f"https://raw.githubusercontent.com/{info.owner}/{info.repo}/{ref}/{path}"
    if info.forge == ForgeType.GITLAB:
        return f"https://gitlab.com/{info.owner}/{info.repo}/-/raw/{ref}/{path}"
    if info.forge == ForgeType.CODEBERG:
        return f"https://codeberg.org/{info.owner}/{info.repo}/raw/branch/{ref}/{path}"
    if info.forge == ForgeType.GITEA:
        return f"https://{info.host}/{info.owner}/{info.repo}/raw/branch/{ref}/{path}"

    return None


async def is_repo_public(repo_url: str) -> bool:
    """Check whether a repository is publicly accessible.

    Does a HEAD request to the repo's web URL. Returns ``True`` if
    the server responds with 200 (public). Returns ``False`` on 404,
    403, or any error.
    """
    url = normalize_clone_url(repo_url)
    try:
        async with httpx.AsyncClient(timeout=10.0, follow_redirects=True) as client:
            resp = await client.head(url)
            return resp.status_code == 200
    except httpx.HTTPError as e:
        _log.debug("is_repo_public check failed", url=url, error=str(e))
        return False


def normalize_clone_url(repo_url: str) -> str:
    """Normalize a repo URL for cloning.

    Strips trailing slashes, ``.git`` suffix, and URL fragments.
    """
    url = repo_url.strip()
    # Remove fragment.
    url = url.split("#")[0]
    # Remove query string.
    url = url.split("?")[0]
    # Remove trailing slashes.
    url = url.rstrip("/")
    # Remove .git suffix (common in clone URLs).
    if url.endswith(".git"):
        url = url[:-4]
    return url

File: crates/nexus-coordinator-rs/src/forge.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Multi-forge URL detection and helpers for verified deploy
//! (Sprint 42 Phase B, port of forge.py S14).

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ForgeType {
    GitHub,
    GitLab,
    Codeberg,
    Gitea,
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ForgeInfo {
    pub forge: ForgeType,
    pub owner: String,
    pub repo: String,
    pub host: String,
}

pub fn normalize_clone_url(repo_url: &str) -> String {
    let mut url = repo_url.trim().to_string();
    if let Some(pos) = url.find('#') {
        url.truncate(pos);
    }
    if let Some(pos) = url.find('?') {
        url.truncate(pos);
    }
    let url = url.trim_end_matches('/');
    let url = url.strip_suffix(".git").unwrap_or(url);
    url.to_string()
}

pub fn detect_forge(repo_url: &str) -> ForgeInfo {
    let url = normalize_clone_url(repo_url);
    let unknown = ForgeInfo {
        forge: ForgeType::Unknown,
        owner: String::new(),
        repo: String::new(),
        host: String::new(),
    };

    let stripped = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
        .unwrap_or_default();

    if stripped.is_empty() {
        return unknown;
    }

    let parts: Vec<&str> = stripped.splitn(4, '/').collect();
    if parts.len() < 3 {
        return unknown;
    }

    let host = parts[0];
    let owner = parts[1].to_string();
    let repo = parts[2].to_string();

    let forge = match host {
        "github.com" => ForgeType::GitHub,
        "gitlab.com" => ForgeType::GitLab,
        "codeberg.org" => ForgeType::Codeberg,
        _ => ForgeType::Gitea,
    };

    ForgeInfo {
        forge,
        owner,
        repo,
        host: host.to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_strips_git_suffix() {
        assert_eq!(
            normalize_clone_url("https://github.com/user/repo.git"),
            "https://github.com/user/repo"
        );
    }

    #[test]
    fn normalize_strips_trailing_slash() {
        assert_eq!(
            normalize_clone_url("https://github.com/user/repo/"),
            "https://github.com/user/repo"
        );
    }

    #[test]
    fn normalize_strips_fragment_and_query() {
        assert_eq!(
            normalize_clone_url("https://github.com/user/repo?tab=code#readme"),
            "https://github.com/user/repo"
        );
    }

    #[test]
    fn detect_github() {
        let info = detect_forge("https://github.com/user/repo.git");
        assert_eq!(info.forge, ForgeType::GitHub);
        assert_eq!(info.owner, "user");
        assert_eq!(info.repo, "repo");
    }

    #[test]
    fn detect_gitlab() {
        let info = detect_forge("https://gitlab.com/org/project");
        assert_eq!(info.forge, ForgeType::GitLab);
        assert_eq!(info.owner, "org");
        assert_eq!(info.repo, "project");
    }

    #[test]
    fn detect_codeberg() {
        let info = detect_forge("https://codeberg.org/user/repo");
        assert_eq!(info.forge, ForgeType::Codeberg);
    }

    #[test]
    fn detect_self_hosted_gitea() {
        let info = detect_forge("https://git.example.com/user/repo");
        assert_eq!(info.forge, ForgeType::Gitea);
        assert_eq!(info.host, "git.example.com");
    }

    #[test]
    fn detect_unknown_no_path() {
        let info = detect_forge("https://example.com");
        assert_eq!(info.forge, ForgeType::Unknown);
    }
}
